### ISSUE TITLE
Advapi32 AccessCheck non-optional parameter

### ISSUE THREAD

**Issue by AliveDevil** [microsoft/win32metadata#1771]
The Advapi32 AccessCheck method is defined as follows in Win32Metadata:
```
public unsafe static extern BOOL AccessCheck(
    [In] PSECURITY_DESCRIPTOR pSecurityDescriptor,
    [In] HANDLE ClientToken,
    [In] uint DesiredAccess,
    [In] GENERIC_MAPPING* GenericMapping,
    [Optional][Out][MemorySize(BytesParamIndex = 5)] PRIVILEGE_SET* PrivilegeSet,
    [In][Out] uint* PrivilegeSetLength,
    [Out] uint* GrantedAccess,
    [Out] BOOL* AccessStatus);
```
With PrivilegeSet and PrivilegeSetLength in C-headers (`securitybaseapi.h`):
```
    _Out_writes_bytes_to_opt_(*PrivilegeSetLength,*PrivilegeSetLength) PPRIVILEGE_SET PrivilegeSet,
    _Inout_ LPDWORD PrivilegeSetLength,
```
As per documentation[^1] the parameter `PrivilegeSet` is optional:
> When a pointer parameter annotation includes _opt_, it indicates that the parameter may be null

On accessing this method through CsWin32 it turned out this parameter is not optional at all.

Assuming optional:
```
// Pass null for both
AccessCheck(…, null, null, …) = FALSE
GetLastError() = 998 (ERROR_NOACCESS)

// Pass null for Set, and 0 for Length
uint size = 0;
AccessCheck(…, null, &size, …) = FALSE
GetLastError() = 122 (ERROR_INSUFFICIENT_BUFFER)

// Pass null for Set, and 20 (minimum size for PRIVILEGE_SET) for Length
uint size = 20;
AccessCheck(…, null, &size, …) = FALSE
GetLastError() = 998 (ERROR_NOACCESS)

// Pass buffer, with zero length:
PRIVILEGE_SET set = default;
uint size = 0;
AccessCheck(…, &set, &size, …) = FALSE
GetLastError() = 122 (ERROR_INSUFFICIENT_BUFFER)
```

PRIVILEGE_SET size is 20.
Reverse test, assuming both parameters are in fact not optional:
```
PRIVILEGE_SET = default;
uint size = sizeof(PRIVILEGE_SET);
AccessCheck(…, &set, &size, …) = TRUE
```

Please let me know whether this is wrongly represented in the C-headers (and could be patched in Win32Metadata), or is an implementation error in AccessCheck.

[^1]: [Annotating function parameters and return values#Optional pointer parameters](https://learn.microsoft.com/cpp/code-quality/annotating-function-parameters-and-return-values#optional-pointer-parameters)

**Comment by mikebattista** [comment]
@jonwis could help route this question to the appropriate API owner? Looks like the metadata matches the headers but the implementation is questionable.

**Comment by riverar** [comment]
I don't think there's an issue here.

Setting the privilege set to `NULL` allows the caller to ask the system what size it needs for the privilege set structure. (This is a common Win32 API pattern.) `PrivilegeSetLength` will receive the size needed, which can then be used in sizing operations and subsequent calls.

Here's a working example, error checking elided:

```cpp
#include <windows.h>
#include <assert.h>
#include <iostream>

int main()
{
    HANDLE primary_token;
    OpenProcessToken(GetCurrentProcess(), TOKEN_ALL_ACCESS, &primary_token);

    HANDLE impersonation_token;
    DuplicateToken(primary_token, SecurityImpersonation, &impersonation_token);

    auto path = LR"(C:\temp\test.txt)";
    DWORD descriptor_size = 0;
    GetFileSecurity(path,
        OWNER_SECURITY_INFORMATION | GROUP_SECURITY_INFORMATION | DACL_SECURITY_INFORMATION, nullptr, 0, &descriptor_size);

    auto descriptor = static_cast<PSECURITY_DESCRIPTOR>(std::malloc(descriptor_size));
    GetFileSecurity(path,
        OWNER_SECURITY_INFORMATION | GROUP_SECURITY_INFORMATION | DACL_SECURITY_INFORMATION, descriptor, descriptor_size, &descriptor_size);

    GENERIC_MAPPING mapping{};
    DWORD granted_access_mask = 0;
    BOOL access_status = FALSE;
    DWORD privilege_set_size = 0;
    PRIVILEGE_SET privilege_set = {};
    
    auto result = AccessCheck(descriptor, impersonation_token, descriptor_size, &mapping, nullptr, &privilege_set_size, &granted_access_mask, &access_status);
    assert(GetLastError() == ERROR_INSUFFICIENT_BUFFER);
    
    result = AccessCheck(descriptor, impersonation_token, descriptor_size, &mapping, &privilege_set, &privilege_set_size, &granted_access_mask, &access_status);
    assert(result == TRUE);
    assert(access_status == TRUE);
}

```

**Comment by AliveDevil** [comment]
Thanks for checking. I recreated the initial set in a MSTest in C#:
https://github.com/AliveDevil/win32metadata-1771

Realized now that I had not seen that
```
DWORD size = 0;
AccessCheck(…, null, &size, …)
```
returned `INSUFFICIENT_BUFFER`.

Nevertheless, my initial point still stands: `[Optional]` annotated parameters shouldn't cause errors, when they are not required/used.

The way I see this: `PrivilegeSet` is optional, thus the `PrivilegeSetLength` parameter is only needed if I want `PrivilegeSet` to have meaning.
On passing `PrivilegeSetLength=nullptr` I'm requesting the `PrivilegeSet`-optional handling - i.e. ignore it, completely.
On passing anything else to PrivilegeSetLength (i.e. `0` or some buffer length) I expect PrivilegeSet to be used (i.e. passing a buffer).

**Comment by mikebattista** [comment]
Can you summarize what change you'd like to see in the metadata?

**Comment by riverar** [comment]
> The way I see this: `PrivilegeSet` is optional, thus the `PrivilegeSetLength` parameter is only needed if I want `PrivilegeSet` to have meaning.
> On passing `PrivilegeSetLength=nullptr` I'm requesting the `PrivilegeSet`-optional handling - i.e. ignore it, completely.
> On passing anything else to PrivilegeSetLength (i.e. `0` or some buffer length) I expect PrivilegeSet to be used (i.e. passing a buffer), if it is then failing due to me passing a `nullptr` and a non-zero length, then that is my fault and expected.

`PrivilegeSetLength` is not optional, so passing `nullptr` there is invalid. It is used to describe the size of the input `PrivilegeSet` or to communicate the required size of the `PrivilegeSet`.

Does that help clarify?

**Comment by mikebattista** [comment]
`PrivilegeSetLength` is already not optional in the metadata.

**Comment by riverar** [comment]
@AliveDevil Submitted a PR to the tests as an alternate explanation, hope that helps! https://github.com/AliveDevil/win32metadata-1771/pull/1

**Comment by AliveDevil** [comment]
Thanks @riverar for taking the time looking into this.

I think the mistake I did in looking at the API was that I saw the optional Ptr and assumed the parameter pair `(Ptr, Length)` is `Optional (Ptr, Length)` not `Required (NotNullIfLengthNotZero Ptr, Length)` - I'd attribute this difference in interpretation to culture/language differences.

> Can you summarize what change you'd like to see in the metadata?

Making clear that the set-ptr parameter is not optional - optional, in my understanding is fundamentally different than "must be set, when associated parameter is non-zero". To be fair, I don't have a good idea how to make this clear - both in the Win32 API, as well as in the metadata project. Maybe not including the optional parameter is sufficient, but as this falls more into the Win32-docs department I do acknowledge that there is nothing to be done in the metadata representation of this.

@mikebattista 


To repeat myself: Thanks for looking into this, I do appreciate the feedback here. :+1:
